### PR TITLE
Build Alpine (musl) version of the CLR Profiler

### PR DESCRIPTION
**Pull request by zacharycmontoya**
Issue
Previously, only one version of the CLR Profiler was built for Linux distributions, and the library linked against glibc as the C standard library. Musl-based distributions (meaning they use musl as the C standard library instead of glibc) then required a combination of additional `libc6-compat` and `gcompat` packages to properly load and run the profiler library. However, on Alpine versions 3.10+ the additional compatibility packages fail and we're left with no support for these distributions.

Changes proposed in this pull request:
- Build `Datadog.Trace.ClrProfiler.Native.so` on an Alpine container so it is linked against musl, and package it as `datadog-dotnet-apm-$VERSION-musl.tar.gz`.
- Create new Alpine dockerfiles for building, testing, and packaging the new version of the library. Note: There are currently no apk packages for .NET Core so there is a separate dockerfile for each version of the .NET Core runtime on Alpine image. 
- For building, testing, and packaging the new musl-based `Datadog.Trace.ClrProfiler.Native.so`, run the Azure DevOps pipelines in separate jobs (Linux virtual machines) from the other Linux builds, so there are no collisions when producing the musl-based `Datadog.Trace.ClrProfiler.Native.so` in CI.
- Modify the IntegrationTests bash file to directly invoke `dotnet vstest` so the command directly tests the test binaries instead of trying to rebuild the test projects, which causes issues on the Alpine images that only have .NET Core SDK installed
- Add a minimal ConsoleApp .NET Core app that shows how to build a Docker container with the automatic instrumentation for a Debian container and various Alpine containers

@DataDog/apm-dotnet

### CODE REVIEW

**Comment by zacharycmontoya** [comment]
For reference, here are two integration-test runs before and after consuming the new package on Alpine images. In the before test, the Alpine images use the two compat packages `libc6-compat` (for all Alpine versions) and `gcompat` (for Alpine 3.9+) that we document on our [.NET Tracer setup](https://docs.datadoghq.com/tracing/setup/dotnet/?tab=netcoreonlinux#installation). You'll notice that the Alpine netcoreapp2.1 run is successful because that container runs on Alpine 3.9, but Alpine netcoreapp3.0 and Alpine netcoreapp3.1 fail because they run on Alpine 3.10. Once they consume the new musl-based library, they pass on both distributions and all runtimes.

Before: https://dev.azure.com/datadog-apm/dd-trace-dotnet/_build/results?buildId=15622&view=ms.vss-test-web.build-test-results-tab

After: https://dev.azure.com/datadog-apm/dd-trace-dotnet/_build/results?buildId=15625&view=ms.vss-test-web.build-test-results-tab

File: docker/Datadog.Trace.ClrProfiler.IntegrationTests.sh [hunk 1]
```diff
@@ -1,20 +1,20 @@
 #!/bin/bash
 set -euxo pipefail
 
-DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null && pwd )"
+cd "$( dirname "${BASH_SOURCE[0]}" )"/../
```
Comment (lucaspimentel) [inline docker/Datadog.Trace.ClrProfiler.IntegrationTests.sh]: Thanks! The need for `..` all over this script always bothered me.

File: customer-samples/ConsoleApp/Alpine3.10.dockerfile [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+COPY . ./
+RUN dotnet publish -c Release -f netcoreapp3.1 -o out
+
+# Build runtime image
+FROM mcr.microsoft.com/dotnet/core/runtime:3.1-alpine3.10
+WORKDIR /app
+COPY --from=build /app/out .
+
+# Set up Datadog APM
+RUN apk --no-cache update && apk add curl
```
Comment (zacharycmontoya) [inline customer-samples/ConsoleApp/Alpine3.10.dockerfile]: Notice here that we don't need any additional packages to run the profiler on Alpine with the new profiler, just `curl` to download the tar.

File: docker/build.sh [hunk 3]
```diff
@@ -44,6 +44,6 @@ done
 
 dotnet msbuild Datadog.Trace.proj -t:RestoreAndBuildSamplesForPackageVersions -p:Configuration=$buildConfiguration -p:ManagedProfilerOutputDirectory="$PUBLISH_OUTPUT" -p:TargetFramework=$publishTargetFramework
 
-for proj in Datadog.Trace.ClrProfiler.IntegrationTests ; do
+for proj in Datadog.Trace.IntegrationTests Datadog.Trace.OpenTracing.IntegrationTests Datadog.Trace.ClrProfiler.IntegrationTests ; do
```
Comment (lucaspimentel) [inline docker/build.sh]: I just noticed this as well while working on #611. We don't [run the other projects in CI](https://github.com/DataDog/dd-trace-dotnet/blob/5e6adfb739de4a55b9f2f2efa3380b6be28072af/.azure-pipelines/integration-tests.yml#L277) either.

File: customer-samples/ConsoleApp/Debian.dockerfile [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+COPY . ./
+RUN dotnet publish -c Release -f netcoreapp3.1 -o out
+
+# Build runtime image
+FROM mcr.microsoft.com/dotnet/core/runtime:3.1
+WORKDIR /app
+COPY --from=build /app/out .
+
+# Set up Datadog APM
+ENV TRACER_VERSION=1.11.0
```
Comment (lucaspimentel) [inline customer-samples/ConsoleApp/Debian.dockerfile]: I think if you don't bump this to `1.12.0` now the version bump tool won't be able to catch it. Do we care?

File: customer-samples/ConsoleApp/README.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# .NET Core Docker Sample
+This sample demonstrates how to build container images for .NET Core apps that use the Datadog .NET Tracer.
+
+The instructions assume that you have cloned this repo, have [Docker](https://www.docker.com/products/docker) installed, and have a command prompt open within the customer-samples/ConsoleApp directory within the repo.
```
Comment (lucaspimentel) [inline customer-samples/ConsoleApp/README.md]:
```suggestion
The instructions assume that you have cloned this repo, have [Docker](https://www.docker.com/products/docker) installed, and have a command prompt open within the `customer-samples/ConsoleApp` directory in the repo.
```

Comment (zacharycmontoya) [reply customer-samples/ConsoleApp/README.md]:
```suggestion
The instructions assume that you have cloned this repo, have [Docker](https://www.docker.com/products/docker) installed, and have a command prompt open in the `customer-samples/ConsoleApp` directory.
```

File: docker/build.sh (hunk 3 above)
Comment (zacharycmontoya) [reply docker/build.sh]: I didn't realize that. I'll address that in this PR

File: customer-samples/ConsoleApp/Debian.dockerfile (hunk 4 above)
Comment (zacharycmontoya) [reply customer-samples/ConsoleApp/Debian.dockerfile]: Yeah it's probably good to keep this up-to-date with the latest release. Fixed in 064ea569ab3248e5e99c3dd32857b41231eefe8f

File: docker/build.sh (hunk 3 above)
Comment (zacharycmontoya) [reply docker/build.sh]: Fixed in fdd7f513704e2e2ed71e6f4f3b206c36a3d527ff

**Pull request merged**